crates/configuration/src/network.rs

```diff
@@ -178,6 +180,97 @@ impl NetworkConfigBuilder<Initial> {
+        }
+
+        // retrieve the defaults relaychain for assigning to nodes if needed
+        let relaychain_default_command = network_config
+            .relaychain
+            .as_ref()
+            .unwrap()
+            .default_command()
+            .cloned();
+
+        let relaychain_default_image = network_config
+            .relaychain
+            .as_ref()
+            .unwrap()
+            .default_image()
+            .cloned();
+
+        let relaychain_default_resources = network_config
+            .relaychain
+            .as_ref()
+            .unwrap()
+            .default_resources()
+            .cloned();
+
+        let relaychain_default_db_snapshot = network_config
+            .relaychain
+            .as_ref()
+            .unwrap()
+            .default_db_snapshot()
+            .cloned();
+
+        let default_args = network_config.relaychain().default_args.clone();
+
+        // if there is some existing defaults and
+        if relaychain_default_command.is_some() {
+            // here we've got our own clone of Option<Command> so we just unwrap it because we asserted above it exists
+            let default_command: Option<crate::shared::types::Command> = relaychain_default_command;
+            // we take all nodes as mutables
+            for node in network_config.relaychain.as_mut().unwrap().nodes.iter_mut() {
+                // we modify only nodes which don't arleady have a command
+                if node.command.is_none() {
+                    node.command = default_command.clone();
+                }
+            }
+        }
+
+        if relaychain_default_image.is_some() {
+            let default_image = relaychain_default_image;
+            for node in network_config.relaychain.as_mut().unwrap().nodes.iter_mut() {
+                if node.image.is_none() {
+                    node.image = default_image.clone();
+                }
+            }
+        }
+
+        if relaychain_default_resources.is_some() {
+            let default_resources = relaychain_default_resources;
+            for node in network_config.relaychain.as_mut().unwrap().nodes.iter_mut() {
+                if node.resources.is_none() {
+                    node.resources = default_resources.clone();
+                }
+            }
+        }
+
+        if relaychain_default_db_snapshot.is_some() {
+            let default_db_snapshot = relaychain_default_db_snapshot;
+            for node in network_config.relaychain.as_mut().unwrap().nodes.iter_mut() {
+                if node.db_snapshot.is_none() {
+                    node.db_snapshot = default_db_snapshot.clone();
+                }
+            }
+        }
+
+        if !default_args.is_empty() {
+            for node in network_config.relaychain.as_mut().unwrap().nodes.iter_mut() {
+                if node.args.is_empty() {
+                    node.args.clone_from(&default_args.clone());
+                }
+            }
+        }
```

Comment:
We are iterating over the `nodes` here multiple times, one for each option. I think we can rewrite in the other way and iterate one time and make the check inside the loop. (I can send you a pr)